src/vivliostyle/repetitiveelements.js

```diff
@@ -0,0 +1,909 @@
+        /** @const */ this.rootSourceNode = rootSourceNode;
+        /** @type {boolean} */ this.doneInitialLayout = false;
+        /** @type {boolean} */ this.isRoot = false;
+        /** @type {vivliostyle.repetitiveelements.RepetitiveElements} */ this.repetitiveElements = null;
+    };
+    /** @const */ var RepetitiveElementsOwnerFormattingContext = vivliostyle.repetitiveelements.RepetitiveElementsOwnerFormattingContext;
+
+    /**
+     * @override
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.getName = function() {
+        return "Repetitive elements owner formatting context (vivliostyle.repetitiveelements.RepetitiveElementsOwnerFormattingContext)";
+    };
+
+    /**
+     * @override
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.isFirstTime = function(nodeContext, firstTime) {
+        return firstTime;
+    };
+
+    /**
+     * @override
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.getParent = function() {
+        return this.parent;
+    };
+
+    /**
+     * @return {vivliostyle.repetitiveelements.RepetitiveElements}
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.getRepetitiveElements = function() {
+        return this.repetitiveElements;
+    };
+
+    /**
+     * @param {adapt.vtree.NodeContext} nodeContext
+     * @return {boolean}
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.isAfterContextOfRootElement = function(nodeContext) {
+        if (nodeContext
+            && nodeContext.formattingContext === this
+            && nodeContext.sourceNode === this.rootSourceNode
+            && nodeContext.after) return true;
+        for (; nodeContext; nodeContext = nodeContext.parent) {
+            if (nodeContext.formattingContext === this) {
+                return false;
+            }
+        }
+        return true;
+    };
+
+    /**
+     * @param {adapt.vtree.NodeContext} position
+     * @returns {Element}
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.getRootViewNode = function(position) {
+        var root = this.getRootNodeContext(position);
+        return root ? /** @type {Element} */ (root.viewNode) : null;
+    };
+
+    /**
+     * @param {adapt.vtree.NodeContext} nodeContext
+     * @returns {adapt.vtree.NodeContext}
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.getRootNodeContext = function(nodeContext) {
+        do {
+            if (!this.isInherited(nodeContext)) {
+                return nodeContext;
+            }
+        } while (nodeContext = nodeContext.parent);
+        return null;
+    };
+
+    /**
+     * @param {adapt.vtree.NodeContext} nodeContext
+     * @returns {boolean}
+     */
+    RepetitiveElementsOwnerFormattingContext.prototype.isInherited = function(nodeContext) {
+        return !!(nodeContext
+          && nodeContext.formattingContext === this
+          && nodeContext.sourceNode !== this.rootSourceNode);
+    };
```

Comment:
I'm a bit uncomfortable with the method name `isInherited`.
My understanding is that the meaning of this method is: "this `nodeContext` belongs to the `formattingContext`".
The "root" node of a formatting context *generates* the formatting context, but does not *belongs* to it.
Therefore I think a method like `nodeContext.belongsTo(formattingContext)` is more natural.
Since a formatting context always has its "root" node, it should be possible to make the method work with any formatting context, not limited to `RepetitiveElementsOwnerFormattingContext`.